lib/Sema/CSSimplify.cpp

```diff
@@ -1091,14 +1091,16 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
                                      ConstraintLocatorBuilder locator) {
   // An @autoclosure function type can be a subtype of a
   // non-@autoclosure function type.
-  if (func1->isAutoClosure() != func2->isAutoClosure() &&
-      kind < ConstraintKind::Subtype)
-    return SolutionKind::Error;
+  if (func1->isAutoClosure() != func2->isAutoClosure()) {
+    assert(func1->isAutoClosure() && "Missing implicit autoclosure wrapper");
+    if (kind < ConstraintKind::Subtype)
+      return SolutionKind::Error;
+  }
   
   // A non-throwing function can be a subtype of a throwing function.
   if (func1->throws() != func2->throws()) {
     // Cannot drop 'throws'.
-    if (func1->throws() || (func2->throws() && kind < ConstraintKind::Subtype))
+    if (func1->throws() || kind < ConstraintKind::Subtype)
```

Comment:
Can you explain why you relaxed this?

Reply:
As I wrote in the commit message, in this scope func1->throws() and func2->throws() are mutually exclusive. If func1->throws() is false, then func2->throws() must be true.